### PR TITLE
SSL4EO-S12: add new dataset/datamodule

### PR DESCRIPTION
**Pull request by adamjstewart**
Adds the [SSL4EO-S12](https://github.com/zhu-xlab/SSL4EO-S12) dataset.

Some things the dataset doesn't currently support because I don't need them, but could potentially be added someday:

* Automatic downloading: same as SEN12MS and So2Sat, we haven't yet figured out automatic downloads from the TUM servers
* RGB version, 100-patch MSI subset, 50K RGB subset: alternate versions of the dataset on OneDrive and Google Drive
* Combining s1, s2c, or s2a: we don't yet have an easy model for supporting this in trainers and transforms
* Spectral band subsets: adds complexity to the dataset
* GeoDataset: each file has geographic metadata, we could build a GeoDataset

Plotting seems broken at the moment, will upload plots once that's working better.

### CODE REVIEW

File: tests/data/ssl4eo/data.py
Comment (adamjstewart) [inline tests/data/ssl4eo/data.py:88]: There's probably a better way to do this...

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:45]: Popcorn not included. BYOP.

Still trying to extract the full dataset, will update XX once I know the full time it takes.

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:141]: This is somewhat hard-coded, but is much faster than storing a list of every directory like we do in SeCo.

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:147]: Each Sentinel band is in its original resolution (10–60 m) and needs to be resampled to the higher resolution.

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:206]: Wish we could call the plot methods of Sentinel1 and Sentinel2 directly so we didn't have to copy the same logic into every dataset. I bet we could figure out a way to do this...

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:215]: @RitwikGupta this plotting trick isn't working, all I see is a black image with red specs. I know you're not the biggest fan of RGB plotting, but this dataset presumably only has one type of image processing so it should be possible to reliably plot. Not sure how to tell if the image is power/decibel/amplitude or gamma/sigma and what that actually means for plotting. Any advice? You can download the [100 patch subset](https://github.com/zhu-xlab/SSL4EO-S12) if you want to test the plotting method.

Comment (RitwikGupta) [reply torchgeo/datasets/ssl4eo.py:215]: @adamjstewart The Sentinel-1 imagery they pulled from GEE is in the decibel scale. Normally, Sentinel-1 data from Copernicus is in the power scale. The ratio you want to calculate should be in the power scale. This code will work to do visualize this specific data:

```
cross_pol = np.exp(cross_pol / 10)
co_pol = np.exp(co_pol / 10)
ratio = cross_pol / co_pol

fig, ax = plt.subplots(1, 4, figsize=(20, 10))
ax[0].imshow(cross_pol, cmap='bone')
ax[1].imshow(co_pol, cmap='bone')
ax[2].imshow(ratio, cmap='bone')
ax[3].imshow(np.concatenate([cross_pol, co_pol, ratio], axis=2))
```
image

Comment (RitwikGupta) [reply torchgeo/datasets/ssl4eo.py:215]: Here's the relevant GEE documentation that says they convert to decibel when pulling S1 data in GEE: https://developers.google.com/earth-engine/datasets/catalog/COPERNICUS_S1_GRD
image

Comment (adamjstewart) [reply torchgeo/datasets/ssl4eo.py:215]: 1. I thought it was supposed to be [co_pol, cross_pol, ratio], not [cross_pol, co_pol, ratio]?
2. Am I not supposed to do [normalization](https://gis.stackexchange.com/a/400780/123758) (co_pol / 0.3, cross_pol / 0.05, ratio / 25)?
3. Even with divide by 10, no normalization, swap co_pol/cross_pol, the images don't look great:
Figure_1

Comment (RitwikGupta) [reply torchgeo/datasets/ssl4eo.py:215]: Which image are you using? I can test with the same one.
* I think you can keep [co_pol, cross_pol, ratio].
* Are you doing `np.exp(x / 10)` or just `x /10`?
* You should use the log scale values in that post for viz purposes only. It's not required, but it may help. Do min-max norm to those ranges, not just division.

**Comment by adamjstewart** [comment]
S1 plotting is definitely still broken, but I would be okay with merging this as is and fixing S1 plotting in a follow-up PR since I need this dataset soon.

I'm planning on adding a datamodule for this, but will wait until #1168 is merged so I can properly test it.

**Review by isaaccorley — DISMISSED** [review]

File: torchgeo/datasets/ssl4eo.py
Comment (nilsleh) [inline torchgeo/datasets/ssl4eo.py:24]: "multimodal" typo

File: torchgeo/datasets/ssl4eo.py
Comment (nilsleh) [inline torchgeo/datasets/ssl4eo.py:102]: As far as I understand "seasons" decides how many seasonal patches to sample and it does so randomly. Would you also want to give control over the seasons in the sense if a user only wants to sample from a particular or a subset of the four seasons?

Comment (adamjstewart) [reply torchgeo/datasets/ssl4eo.py:102]: That could be a feature worth adding later, but I'm trying to keep things simple until someone actually needs that functionality.

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [reply torchgeo/datasets/ssl4eo.py]: Submitted a PR to fix upstream README where I copied this from.

**Review by nilsleh — APPROVED** [review]

**Comment by calebrob6** [comment]
Can you post an example from plot across the three different splits?

**Comment by adamjstewart** [comment]
@wangyi111 do you want to review this? We couldn't figure out a good way to plot the images, but the rest of the data loader should work.

File: torchgeo/datasets/ssl4eo.py
Comment (wangyi111) [reply torchgeo/datasets/ssl4eo.py]: why do you want to use power values to plot? In my paper, I simply plot the dB values (99% percentile and scale each one to [0,1])

**Comment by wangyi111** [comment]
@adamjstewart I think the problem is from `SSL4EO.metadata['s2c']['bands']`. One band 'B10' missing.

**Comment by adamjstewart** [comment]
Good catch! I think I looked at s2a, didn't see B10, and assumed s2c would be the same. Let me fix that.

File: torchgeo/datasets/ssl4eo.py
Comment (wangyi111) [reply torchgeo/datasets/ssl4eo.py]: then can also plot dB for each channel? e.g. [VV,VH,VV/VH]?

Comment (adamjstewart) [reply torchgeo/datasets/ssl4eo.py]: I think that's what I was doing originally and it didn't work either. I just copied the plotting code from our `Sentinel1` dataset, which worked fine for data downloaded from ASF DAAC, but apparently not for data downloaded from GEE.

**Comment by calebrob6** [comment]
I don't have the S1 downloaded so can't check it
image

**Comment by adamjstewart** [comment]
> I don't have the S1 downloaded so can't check it

You can download the 100-patch subset for testing.

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:1]: @calebrob6 this directly contradicts the change made in #1027. Should we revert that PR too? Given that:

1. All of our datasets use a different scale factor, ignoring which satellite their images come from
2. All images are normalized before they are returned from the dataset or datamodule, changing the appropriate scale factor

I'm starting to think we should use one of the default enhancements from QGIS:

1. Min / max
2. 2% / 98%
3. ± 2 std dev

There could even be an option to select which of these you want to use. We can decide and implement this later, but I want to make sure we're not contradicting ourselves and won't revert #1027 multiple times until we come up with a uniform policy. This also relates to #496.

Comment (calebrob6) [reply torchgeo/datasets/ssl4eo.py:1]: dividing by 10k obviously didn't work though

**Comment by calebrob6** [comment]
image

s1 plotting is broken

File: torchgeo/datasets/ssl4eo.py
Comment (adamjstewart) [reply torchgeo/datasets/ssl4eo.py:1]: Then why did we merge #1027?

**Comment by calebrob6** [comment]
I'm guessing the images from gee and other places are different

**Comment by adamjstewart** [comment]
Yep. We could just plot 2 grayscale images instead of one RGB image for now and fix this in the future when someone (me) has a better understanding of SAR imagery. I'd prefer not to merge this with broken plotting, but we could open an issue and promise to fix it before 0.5.0 is released.

**Comment by RitwikGupta** [comment]
Sorry all. I was out on vacation and then sick for a bit. I will take a look at this again. GEE imagery is in the decibel scale instead of the power scale most Sentinel-1 imagery is released in. This is standard for analysis and visualization.

**Comment by adamjstewart** [comment]
Any updates @RitwikGupta?

We might want to just merge this PR and fix S1 plotting at a later date.

**Comment by calebrob6** [comment]
Yep, I would vote to make the plot show _something_ and merge

**Comment by adamjstewart** [comment]
Alright, see how this looks:

Figure_1
Figure_2
Figure_3
Figure_4

**Review by calebrob6 — APPROVED** [review]

**Pull request merged**
